### PR TITLE
#159987728 Users can make reports on articles

### PR DESCRIPTION
**Pull request by emekafredy**
# What does this PR do?
Create endpoints for users to report articles and for admin to work on these reports

# Description of Task to be completed?
### Create endpoints:
- Report article ``` POST api/articles/<article_slug>/report ```
- View all reports ``` GET api/admin/articles/reports ```
- View a single report ``` GET api/admin/articles/reports/<reportId>```
- View all reports for a single article ``` GET api/admin/reports/articles/<article_slug> ```
- Block article ``` PUT api/admin/articles/:article_slug/delist ```
- View blocked articles ``` GET api/admin/articles/delisted ```
- View a blocked article ``` GET api/admin/articles/:article_slug/delisted ```

- Unblock article ``` PUT api/admin/articles/:article_slug/relist ```
### Update swagger documentation for the above endpoints

# How should this be manually tested?
- Clone the repo
- cd into the 'Thor-ah'
- Run 'npm install'
- Use the .env.sample format to create a .env file
- Run 'npm run migrate'
- Run 'npm run start:dev'

# What are the relevant pivotal tracker stories?
[159987728](https://www.pivotaltracker.com/story/show/159987728)

### CODE REVIEW

File: server/controllers/adminHandleReports.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        },
+        include: [{
+          model: User,
+          as: 'author',
+          attributes: ['username', 'email', 'bio', 'image']
+        }]
+      }]
+    })
+      .then(report => res.status(200).json({
+        report
```
Comment (CEOehis) [inline server/controllers/adminHandleReports.js]: You can add a `status` property with value `success` in your response. Same with line `99-102`

**Review by CEOehis — APPROVED** [review]

**Review by CEOehis — CHANGES_REQUESTED** [review]
You can add a `status` property with value `success` in your response in lines `67-69`. Same thing with line `99-102`

File: server/controllers/adminHandleReports.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        },
+        include: [{
+          model: User,
+          as: 'author',
+          attributes: ['username', 'email', 'bio', 'image']
+        }]
+      }]
+    })
+      .then(reports => res.status(200).json({
+        reports
```
Comment (emmaadesile) [inline server/controllers/adminHandleReports.js]: You could modify the response as
```
res.status(200).json({
  status: 'success',
  report,
})
```

File: server/controllers/adminHandleReports.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+          exclude: ['userId', 'createdAt', 'updatedAt']
+        },
+      }, {
+        model: User,
+        as: 'author',
+        attributes: ['username', 'email', 'bio', 'image']
+      }]
+    })
+      .then(report => res.status(200).json({
+        report
```
Comment (emmaadesile) [inline server/controllers/adminHandleReports.js]: Same as line `36, 68`

File: server/controllers/article.js [hunk 4]
```diff
@@ -346,6 +361,69 @@ class ArticleController {
...
+        }
+        return ReportsOnArticle.findOne({
+          where: { userId }
+        })
+          .then((found) => {
+            if (found) {
+              return res.status(400).json({
+                status: 'error',
+                error: {
+                  message: 'We already have your report on this article'
```
Comment (emmaadesile) [inline server/controllers/article.js]: A user should be able to report an article more than once. Line `398-410`

**Review by emmaadesile — CHANGES_REQUESTED** [review]

File: server/controllers/article.js [hunk 5]
```diff
@@ -346,6 +361,69 @@ class ArticleController {
...
+    }
+
+    Article.findOne({
+      where: { slug }
+    })
+      .then((article) => {
+        if (!article) {
+          return res.status(404).json({
+            status: 'error',
+            errors: {
```
Comment (d-beloved) [inline server/controllers/article.js]: please change this to `error`

**Review by d-beloved — CHANGES_REQUESTED** [review]

File: server/controllers/article.js (hunk 5 above)
Comment (emekafredy) [reply server/controllers/article.js]: Okay. Will do that immediately

File: server/controllers/article.js (hunk 4 above)
Comment (emekafredy) [reply server/controllers/article.js]: Okay. I guess that makes more sense

File: server/controllers/adminHandleReports.js (hunk 1 above)
Comment (emekafredy) [reply server/controllers/adminHandleReports.js]: Okay. Thanks for noticing this omission

**Review by emmaadesile — APPROVED** [review]
LGTM

**Review by CEOehis — APPROVED** [review]
LGTM

File: server/controllers/adminHandleReports.js [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  /**
+   * @description get all reports made by users on all articles
+   * @static
+   * @param {object} req express request object
+   * @param {object} res express response object
+   * @param {object} next next middleware
+   * @returns {next} next calls next
+   * @memberof HandleReports
+   */
+  static getAllReports(req, res, next) {
```
Comment (aknwosu) [inline server/controllers/adminHandleReports.js]: static getReportedArticles(req, res, next) {

File: server/controllers/adminHandleReports.js [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    })
+      .then(reports => res.status(200).json({
+        status: 'success',
+        reports
+      }))
+      .catch(next);
+  }
+
+  /**
+   * @description get a single report made by a user on an article
```
Comment (aknwosu) [inline server/controllers/adminHandleReports.js]: * @description get a single reported article

File: server/controllers/adminHandleReports.js [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+   * @static
+   * @param {object} req express request object
+   * @param {object} res express response object
+   * @param {object} next next middleware
+   * @returns {next} next calls next
+   * @memberof HandleReports
+   */
+  static getSingleReport(req, res, next) {
```
Comment (aknwosu) [inline server/controllers/adminHandleReports.js]: getAReportedArticle

File: server/controllers/adminHandleReports.js [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+  /**
+   * @description get a single report made by a user on an article
+   * @static
+   * @param {object} req express request object
+   * @param {object} res express response object
+   * @param {object} next next middleware
+   * @returns {next} next calls next
+   * @memberof HandleReports
+   */
+  static getReportsForSingleArticle(req, res, next) {
```
Comment (aknwosu) [inline server/controllers/adminHandleReports.js]: getReportsForAnArticle

**Review by aknwosu — CHANGES_REQUESTED** [review]

**Review by d-beloved — APPROVED** [review]

**Pull request merged**